File: pkg/kv/mpp.go
// Copyright 2020 PingCAP, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package kv

import (
	"context"
	"strconv"
	"strings"
	"time"

	"github.com/pingcap/kvproto/pkg/mpp"
	"github.com/pingcap/tidb/pkg/util/tiflash"
	"github.com/pingcap/tidb/pkg/util/tiflashcompute"
	"github.com/pingcap/tipb/go-tipb"
	"github.com/tikv/client-go/v2/tikv"
	"github.com/tikv/client-go/v2/tikvrpc"
)

// MppVersion indicates the mpp-version used to build mpp plan
type MppVersion int64

const (
	// MppVersionV0 supports TiFlash version [~, ~]
	MppVersionV0 MppVersion = iota

	// MppVersionV1 supports TiFlash version [v6.6.x, ~]
	MppVersionV1

	// MppVersionV2 supports TiFlash version [v7.3, ~], support ReportMPPTaskStatus service
	MppVersionV2
	// MppVersionV3

	mppVersionMax

	newestMppVersion MppVersion = mppVersionMax - 1

	// MppVersionUnspecified means the illegal or unspecified version, it only used in TiDB.
	MppVersionUnspecified MppVersion = -1

	// MppVersionUnspecifiedName denotes name of UNSPECIFIED mpp version
	MppVersionUnspecifiedName string = "UNSPECIFIED"
)

// ToInt64 transforms MppVersion to int64
func (v MppVersion) ToInt64() int64 {
	return int64(v)
}

// ToMppVersion transforms string to MppVersion
func ToMppVersion(name string) (MppVersion, bool) {
	name = strings.ToUpper(name)
	if name == MppVersionUnspecifiedName {
		return MppVersionUnspecified, true
	}
	v, err := strconv.ParseInt(name, 10, 64)
	if err != nil {
		return MppVersionUnspecified, false
	}
	version := MppVersion(v)
	if version >= MppVersionUnspecified && version <= newestMppVersion {
		return version, true
	}
	return MppVersionUnspecified, false
}

// GetNewestMppVersion returns the mpp-version can be used in mpp plan
func GetNewestMppVersion() MppVersion {
	return newestMppVersion
}

// MPPTaskMeta means the meta info such as location of a mpp task.
type MPPTaskMeta interface {
	// GetAddress indicates which node this task should execute on.
	GetAddress() string
}

// MPPQueryID means the global unique id of a mpp query.
type MPPQueryID struct {
	QueryTs      uint64 // timestamp of query execution, used for TiFlash minTSO schedule
	LocalQueryID uint64 // unique mpp query id in local tidb memory.
	ServerID     uint64
}

// MPPTask means the minimum execution unit of a mpp computation job.
type MPPTask struct {
	Meta         MPPTaskMeta // on which store this task will execute
	ID           int64       // mppTaskID
	StartTs      uint64
	GatherID     uint64
	MppQueryID   MPPQueryID
	TableID      int64      // physical table id
	MppVersion   MppVersion // mpp version
	SessionID    uint64
	SessionAlias string

	PartitionTableIDs  []int64
	TiFlashStaticPrune bool
}

// ToPB generates the pb structure.
func (t *MPPTask) ToPB() *mpp.TaskMeta {
	meta := &mpp.TaskMeta{
		StartTs:         t.StartTs,
		GatherId:        t.GatherID,
		QueryTs:         t.MppQueryID.QueryTs,
		LocalQueryId:    t.MppQueryID.LocalQueryID,
		ServerId:        t.MppQueryID.ServerID,
		TaskId:          t.ID,
		MppVersion:      t.MppVersion.ToInt64(),
		ConnectionId:    t.SessionID,
		ConnectionAlias: t.SessionAlias,
	}
	if t.ID != -1 {
		meta.Address = t.Meta.GetAddress()
	}
	return meta
}

// MppTaskStates denotes the state of mpp tasks
type MppTaskStates uint8

const (
	// MppTaskReady means the task is ready
	MppTaskReady MppTaskStates = iota
	// MppTaskRunning means the task is running
	MppTaskRunning
	// MppTaskCancelled means the task is cancelled
	MppTaskCancelled
	// MppTaskDone means the task is done
	MppTaskDone
)

// MPPDispatchRequest stands for a dispatching task.
type MPPDispatchRequest struct {
	Data    []byte      // data encodes the dag coprocessor request.
	Meta    MPPTaskMeta // mpp store is the location of tiflash store.
	IsRoot  bool        // root task returns data to tidb directly.
	Timeout uint64      // If task is assigned but doesn't receive a connect request during timeout, the task should be destroyed.
	// SchemaVer is for any schema-ful storage (like tiflash) to validate schema correctness if necessary.
	SchemaVar              int64
	StartTs                uint64
	MppQueryID             MPPQueryID
	GatherID               uint64
	ID                     int64 // identify a single task
	MppVersion             MppVersion
	CoordinatorAddress     string
	ReportExecutionSummary bool
	State                  MppTaskStates
	ResourceGroupName      string
	ConnectionID           uint64
	ConnectionAlias        string
}

// CancelMPPTasksParam represents parameter for MPPClient's CancelMPPTasks
type CancelMPPTasksParam struct {
	StoreAddr map[string]bool
	Reqs      []*MPPDispatchRequest
}

// EstablishMPPConnsParam represents parameter for MPPClient's EstablishMPPConns
type EstablishMPPConnsParam struct {
	Ctx      context.Context
	Req      *MPPDispatchRequest
	TaskMeta *mpp.TaskMeta
}

// DispatchMPPTaskParam represents parameter for MPPClient's DispatchMPPTask
type DispatchMPPTaskParam struct {
	Ctx                        context.Context
	Req                        *MPPDispatchRequest
	EnableCollectExecutionInfo bool
	Bo                         *tikv.Backoffer
}

// MPPClient accepts and processes mpp requests.
type MPPClient interface {
	// ConstructMPPTasks schedules task for a plan fragment.
	// TODO:: This interface will be refined after we support more executors.
	ConstructMPPTasks(context.Context, *MPPBuildTasksRequest, time.Duration, tiflashcompute.DispatchPolicy, tiflash.ReplicaRead, func(error)) ([]MPPTaskMeta, error)

	// DispatchMPPTask dispatch mpp task, and returns valid response when retry = false and err is nil.
	DispatchMPPTask(DispatchMPPTaskParam) (resp *mpp.DispatchTaskResponse, retry bool, err error)

	// EstablishMPPConns build a mpp connection to receive data, return valid response when err is nil.
	EstablishMPPConns(EstablishMPPConnsParam) (*tikvrpc.MPPStreamResponse, error)

	// CancelMPPTasks cancels mpp tasks.
	CancelMPPTasks(CancelMPPTasksParam)

	// CheckVisibility checks if it is safe to read using given ts.
	CheckVisibility(startTime uint64) error

	// GetMPPStoreCount returns number of TiFlash stores if there is no error, else return (0, error).
	GetMPPStoreCount() (int, error)
}

// ReportStatusRequest wraps mpp ReportStatusRequest
type ReportStatusRequest struct {
	Request *mpp.ReportTaskStatusRequest
}

// MppCoordinator describes the basic api for executing mpp physical plan.
type MppCoordinator interface {
	// Execute generates and executes mpp tasks for mpp physical plan.
	Execute(ctx context.Context) (Response, []KeyRange, error)
	// Next returns next data
	Next(ctx context.Context) (ResultSubset, error)
	// ReportStatus report task execution info to coordinator
	// It shouldn't change any state outside coordinator itself, since the query which generated the coordinator may not exist
	ReportStatus(info ReportStatusRequest) error
	// Close and release the used resources.
	Close() error
	// IsClosed returns whether mpp coordinator is closed or not
	IsClosed() bool
	// GetComputationCnt returns the number of node cnt that involved in the MPP computation.
	GetNodeCnt() int
}

// MPPBuildTasksRequest request the stores allocation for a mpp plan fragment.
// However, the request doesn't contain the particular plan, because only key ranges take effect on the location assignment.
type MPPBuildTasksRequest struct {
	KeyRanges []KeyRange
	StartTS   uint64

	PartitionIDAndRanges []PartitionIDAndRanges
}

// ExchangeCompressionMode means the compress method used in exchange operator
type ExchangeCompressionMode int

const (
	// ExchangeCompressionModeNONE indicates no compression
	ExchangeCompressionModeNONE ExchangeCompressionMode = iota
	// ExchangeCompressionModeFast indicates fast compression/decompression speed, compression ratio is lower than HC mode
	ExchangeCompressionModeFast
	// ExchangeCompressionModeHC indicates high compression (HC) ratio mode
	ExchangeCompressionModeHC
	// ExchangeCompressionModeUnspecified indicates unspecified compress method, let TiDB choose one
	ExchangeCompressionModeUnspecified

	// RecommendedExchangeCompressionMode indicates recommended compression mode
	RecommendedExchangeCompressionMode ExchangeCompressionMode = ExchangeCompressionModeFast

	exchangeCompressionModeUnspecifiedName string = "UNSPECIFIED"
)

// Name returns the name of ExchangeCompressionMode
func (t ExchangeCompressionMode) Name() string {
	if t == ExchangeCompressionModeUnspecified {
		return exchangeCompressionModeUnspecifiedName
	}
	return t.ToTipbCompressionMode().String()
}

// ToExchangeCompressionMode returns the ExchangeCompressionMode from name
func ToExchangeCompressionMode(name string) (ExchangeCompressionMode, bool) {
	name = strings.ToUpper(name)
	if name == exchangeCompressionModeUnspecifiedName {
		return ExchangeCompressionModeUnspecified, true
	}
	value, ok := tipb.CompressionMode_value[name]
	if ok {
		return ExchangeCompressionMode(value), true
	}
	return ExchangeCompressionModeNONE, false
}

// ToTipbCompressionMode returns tipb.CompressionMode from kv.ExchangeCompressionMode
func (t ExchangeCompressionMode) ToTipbCompressionMode() tipb.CompressionMode {
	switch t {
	case ExchangeCompressionModeNONE:
		return tipb.CompressionMode_NONE
	case ExchangeCompressionModeFast:
		return tipb.CompressionMode_FAST
	case ExchangeCompressionModeHC:
		return tipb.CompressionMode_HIGH_COMPRESSION
	}
	return tipb.CompressionMode_NONE
}

File: rust/crates/tidb-txnkv/src/mpp.rs
// Copyright 2026 PingCAP, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

//! MPP metadata and client/coordinator contracts from `pkg/kv/mpp.go`.

use std::collections::HashMap;
use std::fmt::Write;
use std::time::Duration;

use crate::{KeyRange, PartitionIdAndRanges};
pub use tidb_proto::MppTaskMeta;

/// MPP plan protocol version.
#[derive(Clone, Copy, Debug, Default, Eq, Ord, PartialEq, PartialOrd)]
pub struct MppVersion(i64);

#[allow(non_upper_case_globals)]
impl MppVersion {
    /// Initial protocol.
    pub const V0: Self = Self(0);
    /// TiFlash 6.6+.
    pub const V1: Self = Self(1);
    /// TiFlash 7.3+ task-status reporting.
    pub const V2: Self = Self(2);
    /// TiFlash 9.0+ string serde.
    pub const V3: Self = Self(3);
    /// Illegal or unspecified version.
    pub const Unspecified: Self = Self(-1);
    /// Newest supported version.
    pub const NEWEST: Self = Self::V3;
    /// Source spelling of the unspecified version.
    pub const UNSPECIFIED_NAME: &'static str = "UNSPECIFIED";

    /// Returns the wire integer.
    #[must_use]
    pub const fn as_i64(self) -> i64 {
        self.0
    }

    /// Preserves every raw Go `int64` value carried by an MPP request.
    #[must_use]
    pub const fn from_raw(value: i64) -> Self {
        Self(value)
    }

    /// Parses the source case-insensitive name/integer domain.
    #[must_use]
    pub fn parse(name: &str) -> Option<Self> {
        if name.eq_ignore_ascii_case(Self::UNSPECIFIED_NAME) {
            return Some(Self::Unspecified);
        }
        match name.parse::<i64>().ok()? {
            -1 => Some(Self::Unspecified),
            0 => Some(Self::V0),
            1 => Some(Self::V1),
            2 => Some(Self::V2),
            3 => Some(Self::V3),
            _ => None,
        }
    }
}

/// Physical location of an MPP task.
pub trait MppTaskLocation {
    /// TiFlash address.
    fn address(&self) -> &str;
}

/// Query-wide MPP identity.
#[derive(Clone, Copy, Debug, Default, Eq, PartialEq)]
pub struct MppQueryId {
    /// Query execution timestamp.
    pub query_ts: u64,
    /// TiDB-local query ID.
    pub local_query_id: u64,
    /// TiDB server ID.
    pub server_id: u64,
}

impl MppTaskLocation for MppTaskMeta {
    fn address(&self) -> &str {
        &self.address
    }
}

/// Minimum execution unit of an MPP computation.
pub struct MppTask<L> {
    /// TiFlash location.
    pub location: L,
    /// Task ID.
    pub id: i64,
    /// Start timestamp.
    pub start_ts: u64,
    /// Gather operator ID.
    pub gather_id: u64,
    /// Query identity.
    pub query_id: MppQueryId,
    /// Physical table ID.
    pub table_id: i64,
    /// MPP version.
    pub version: MppVersion,
    /// Session ID.
    pub session_id: u64,
    /// Session alias.
    pub session_alias: String,
    /// Physical partition table IDs.
    pub partition_table_ids: Vec<i64>,
    /// Static-pruning marker.
    pub tiflash_static_prune: bool,
}

impl<L: MppTaskLocation> MppTask<L> {
    /// Produces the exact task-meta field projection.
    #[must_use]
    pub fn to_meta(&self) -> MppTaskMeta {
        MppTaskMeta {
            start_ts: self.start_ts,
            gather_id: self.gather_id,
            query_ts: self.query_id.query_ts,
            local_query_id: self.query_id.local_query_id,
            server_id: self.query_id.server_id,
            task_id: self.id,
            mpp_version: self.version.as_i64(),
            connection_id: self.session_id,
            connection_alias: self.session_alias.clone(),
            address: if self.id == -1 {
                String::new()
            } else {
                self.location.address().to_owned()
            },
            ..MppTaskMeta::default()
        }
    }
}

/// MPP task lifecycle state.
#[derive(Clone, Copy, Debug, Default, Eq, PartialEq)]
#[repr(u8)]
pub enum MppTaskState {
    /// Ready for dispatch.
    #[default]
    Ready = 0,
    /// Running.
    Running,
    /// Cancelled.
    Cancelled,
    /// Finished.
    Done,
}

/// One MPP task dispatch request.
#[derive(Clone, Debug, Default, Eq, PartialEq)]
pub struct MppDispatchRequest {
    /// Encoded DAG request.
    pub data: Vec<u8>,
    /// TiFlash task location.
    pub meta: MppTaskMeta,
    /// Root-task marker.
    pub root: bool,
    /// Connection timeout.
    pub timeout: u64,
    /// Schema version.
    pub schema_version: i64,
    /// Start timestamp.
    pub start_ts: u64,
    /// Query identity.
    pub query_id: MppQueryId,
    /// Gather operator ID.
    pub gather_id: u64,
    /// Task ID.
    pub id: i64,
    /// MPP version.
    pub version: MppVersion,
    /// Coordinator address.
    pub coordinator_address: String,
    /// Execution-summary marker.
    pub report_execution_summary: bool,
    /// Current lifecycle state.
    pub state: MppTaskState,
    /// Resource-group name.
    pub resource_group_name: String,
    /// Connection ID.
    pub connection_id: u64,
    /// Connection alias.
    pub connection_alias: String,
    /// SQL digest.
    pub sql_digest: String,
    /// Plan digest.
    pub plan_digest: String,
}

/// Parameters for cancelling MPP tasks.
#[derive(Clone, Debug, Default, PartialEq)]
pub struct CancelMppTasksParam {
    /// TiFlash addresses selected for cancellation.
    pub store_addresses: HashMap<String, bool>,
    /// Requests whose tasks must be cancelled.
    pub requests: Vec<MppDispatchRequest>,
}

/// Parameters for establishing one MPP stream.
#[derive(Clone, Debug, PartialEq)]
pub struct EstablishMppConnsParam<C, B> {
    /// Request context/cancellation owner.
    pub context: C,
    /// Dispatch request.
    pub request: MppDispatchRequest,
    /// Target task metadata.
    pub task_meta: MppTaskMeta,
    /// Retry backoffer.
    pub backoffer: B,
}

/// Parameters for dispatching one MPP task.
#[derive(Clone, Debug, PartialEq)]
pub struct DispatchMppTaskParam<C, B> {
    /// Request context/cancellation owner.
    pub context: C,
    /// Dispatch request.
    pub request: MppDispatchRequest,
    /// Whether execution information is collected.
    pub collect_execution_info: bool,
    /// Retry backoffer.
    pub backoffer: B,
}

/// MPP store client contract.
pub trait MppClient {
    /// Request context type.
    type Context;
    /// Error type.
    type Error;
    /// Backoffer type.
    type Backoffer;
    /// Dispatch policy.
    type DispatchPolicy;
    /// TiFlash replica-read policy.
    type ReplicaRead;
    /// Dispatch response.
    type DispatchResponse;
    /// Established stream response.
    type StreamResponse;

    /// Schedules task locations for one plan fragment.
    fn construct_mpp_tasks(
        &mut self,
        context: &Self::Context,
        request: &MppBuildTasksRequest,
        timeout: Duration,
        policy: Self::DispatchPolicy,
        replica_read: Self::ReplicaRead,
        on_error: &mut dyn FnMut(&Self::Error),
    ) -> Result<Vec<MppTaskMeta>, Self::Error>;

    /// Dispatches one task; `retry` is true only when the response is unusable.
    fn dispatch_mpp_task(
        &mut self,
        parameters: DispatchMppTaskParam<Self::Context, Self::Backoffer>,
    ) -> Result<(Self::DispatchResponse, bool), Self::Error>;

    /// Establishes one MPP connection.
    fn establish_mpp_connections(
        &mut self,
        parameters: EstablishMppConnsParam<Self::Context, Self::Backoffer>,
    ) -> Result<(Self::StreamResponse, bool), Self::Error>;

    /// Cancels the selected tasks.
    fn cancel_mpp_tasks(&mut self, parameters: CancelMppTasksParam);

    /// Checks whether `start_ts` is visible.
    fn check_visibility(&self, start_ts: u64) -> Result<(), Self::Error>;

    /// Returns the number of TiFlash stores.
    fn mpp_store_count(&self) -> Result<usize, Self::Error>;
}

/// MPP execution coordinator contract.
pub trait MppCoordinator {
    /// Request context type.
    type Context;
    /// Error type.
    type Error;
    /// Initial response type.
    type Response;
    /// Incremental result type.
    type ResultSubset;

    /// Builds and executes all tasks for a physical MPP plan.
    fn execute(
        &mut self,
        context: &Self::Context,
    ) -> Result<(Self::Response, Vec<KeyRange>), Self::Error>;

    /// Returns the next result subset.
    fn next(&mut self, context: &Self::Context) -> Result<Option<Self::ResultSubset>, Self::Error>;

    /// Reports TiFlash task execution status.
    fn report_status(
        &mut self,
        request: tidb_proto::MppReportTaskStatusRequest,
    ) -> Result<(), Self::Error>;

    /// Closes and releases resources.
    fn close(&mut self) -> Result<(), Self::Error>;

    /// Returns whether the coordinator is closed.
    fn is_closed(&self) -> bool;

    /// Returns the number of nodes participating in computation.
    fn node_count(&self) -> usize;
}

/// MPP task-allocation request.
#[derive(Clone, Debug, Default, Eq, PartialEq)]
pub struct MppBuildTasksRequest {
    /// Non-partitioned key ranges; `Some(empty)` remains non-partitioned.
    pub key_ranges: Option<Vec<KeyRange>>,
    /// Start timestamp.
    pub start_ts: u64,
    /// Partitioned key ranges.
    pub partition_id_and_ranges: Vec<PartitionIdAndRanges>,
}

impl MppBuildTasksRequest {
    /// Returns the exact cache-key string used by Go.
    #[must_use]
    pub fn cache_key(&self) -> String {
        let mut output = String::new();
        if let Some(ranges) = self.key_ranges.as_ref() {
            for (index, range) in ranges.iter().enumerate() {
                let _ = write!(
                    output,
                    "range_id{index}{}{}",
                    range.start_key, range.end_key
                );
            }
            return output;
        }
        for partition in &self.partition_id_and_ranges {
            let _ = write!(output, "partition_id{}", partition.id);
            for (index, range) in partition.key_ranges.iter().enumerate() {
                let _ = write!(
                    output,
                    "range_id{index}{}{}",
                    range.start_key, range.end_key
                );
            }
        }
        output
    }
}
